gms/gossiper.cc

```diff
@@ -1300,6 +1300,23 @@ future<> gossiper::do_gossip_to_unreachable_member(gossip_digest_syn message) {
                     addrs.insert(x.first);
                 }
             }
+            // The unreachable endpoints may be reachable via seed
+            // nodes if their IP addresses have changed. Try them as well.
+            // Use get_seeds() since _seeds state is curated by
+            // remove_endpoint() and is_gossip_only_member().
+            auto all_seeds = get_seeds();
+            for (auto&& x : all_seeds) {
+                if (_unreachable_endpoints.contains(x)) {
```

Comment:
Better to check `std::set<inet_address> addrs;` instead of _unreachable_endpoints. The addrs is the one contains the nodes we are going to talk with. 

We do not even need this check at all since addr is a set, the insert below is a no-op if the node is already in addrs.

Reply:
Please let me explain hte purpose of this code.
The contents of the seed file may represent a long gone node or a node with a new IP address. In most cases it is going to represent irrelevant/useless IPs, not the nodes with a different IP address that we failed to reach.

How can we identify these cases? First of all, if all nodes are alive and healthy, we don't need to bother with looking at the seeds. If some node is down, we study the contents of the seed file. 

If it is a node with STATUS=LEFT we don't need to talk with it as well. This is why the code prefers unreachable_endpoints to addrs, addrs is a subset of unreachable_endpoints. Basically, if a seed is already a member of unreachable_endpoints, the main line is going to take care of it, no need to special case it.

Then, if a node is alive, no need to talk with it as well.

So in my view your comment requests me to pessimize the code, and, again, begin pinging irrelevant nodes in some cases.

My meta request to your comment remains the same: what is the benefit of your suggestion? We need to skip alive nodes anyway in the next line, what value is there with changing one set with another? Could you please provide more elaborate rationale for your suggestions, because I fail to see the point of a short note like "The addrs is the one contains the nodes we're going to talk with" (i.e. ok, it does, but so what, why does it make your suggestion valid, I fail to understand).

Reply:
My  first suggestion is that we do not need to look at _unreachable_endpoints again in the seed loop. The previous loop has processed _unreachable_endpoints already.  The all_seeds is now a subset of _unreachable_endpoints. . 

```
1296             std::set<inet_address> addrs;
1297             for (auto&& x : _unreachable_endpoints) {
1298                 // Ignore the node which is decommissioned
1299                 if (get_gossip_status(x.first) != sstring(versioned_value::STATUS_LEFT)) {
1300                     addrs.insert(x.first);
1301                 }
1302             }
1303             // The unreachable endpoints may be reachable via seed
1304             // nodes if their IP addresses have changed. Try them as well.
1305             // Use get_seeds() since _seeds state is curated by
1306             // remove_endpoint() and is_gossip_only_member().
1307             auto all_seeds = get_seeds();
1308             for (auto&& x : all_seeds) {
1309                 if (_unreachable_endpoints.contains(x)) {
1310                     // We are already contacting this endpoint
1311                     continue;
1312                 }
1313                 auto es = get_endpoint_state_for_endpoint_ptr(x);
1314                 if (es && es->is_alive()) {
1315                     // The endpoint is already live
1316                     continue;
1317                 }
1318                 addrs.insert(x);
1319             }
```

For example

_unreachable_endpoints = n1, n2(with LEFT status) ,n3 

after the loop at 1297, we have

addrs = n1, n3

Suppose,  all_seeds = n2, n5

With your code at line 1309, you will skip to talk to n2. But ip n2 could be a new node reusing the ip address. We should talk to n2 since it is in seed list. Do you see the problem?

My second suggestion was 

```
We do not even need this check at all since addr is a set, the insert below is a no-op if the node is already in addrs.
```

So the following code is enough. 

```
 auto all_seeds = get_seeds();
 for (auto&& x : all_seeds) {
     auto es = get_endpoint_state_for_endpoint_ptr(x);
     if (es && es->is_alive()) {
         // The endpoint is already live
         continue;
     }
     addrs.insert(x);
 }       
```

Reply:
No, we should not talk to n2 in your example. n2 can not be a useful seed because it is a *known* node. A node with a new IP address can not be a known node, it cannot have status=LEFT, etc, etc.
We only need to talk to the seeds we're *not* seeing in any other list.

Reply:
@asias ping